utilities/include/vec4.hpp

```diff
@@ -37,28 +37,34 @@ class alignas(ALIGN_WIDTH) vec4{
             return m_v[2];
         }
 
-        constexpr T a() const{
+        constexpr T w() const{
             return m_v[3];
         }
         
+        //indexing operation
+        constexpr T const & operator[](std::size_t i ) const{
```

Comment:
`constexpr` implies `const`, so this function signature is redundant. This should be `constexpr T& operator[](std::size_t i} const`.

You should also include the `cstddef` header for the `std::size_t` type.

You also want to have the `constexpr T operator[](std::size_t i)` getter, so that you can also use it for assignment (like `vec4[3] = 5;`).

Reply:
should the class include both of these( with and without const)?

Reply:
Yes. To be clear, you want both `constexpr T& operator[](std::size_t i} const` and `constexpr T operator[](std::size_t i)`.